### PR TITLE
[CIS-164] Make it possible to set current user

### PR DESCRIPTION
**Pull request by VojtaStavik**
### In this PR:

This PR can also be named **The Great Consolidation**. Finally, all pieces from the "set up groundwork" tasks fit together. 

**Changes:**

- `WebSocketClient` uses mutable `Endpoint` for its WS request. This is needed because of the user switching.

- `CurrentUserModel` subclass is created. It extends `UserModel` with fields valid only for the currently logged-in user.

- `TokenProvider` added to `ChatClientConfig`

- `RequestEncoder` functionality is now extended and it automatically adds the current user's token to requests. The token is obtained the same way the connection Id is obtained (delegate). `Client` sets itself as a delegate for the encoders it creates. This way, the refresh of the token is handled automatically. #magic

- Both `APIClient` and `WebSocketClient` now have (almost) unified API. They both require:
  - `URLSesisonConfiguration` with the common headers (like the app environment, OS, etc...)
  - `RequestEncoder` - `APIClient` uses it to encode all requests, `WebSocketClient` uses it to create the initial connect request.

- Changes to `ChatClient`:
  - The initial user of `Client` is anonymous and the client is not connected by default
  - You can call `connect` and `disconnect` on `Client`
  - It's possible to set a new user
  - It's possible to set a new anonymous user

---

### ⚠️ Important

**Setting a new (anonymous) user has the following consequences:**

- `disconnect` on `WebSocketClient` is called
- the current `connectionId` is reset
- all API calls waiting for `connectionId` are cancelled
- the id of the new user is set to `currentUserId`
- the connect endpoint of `WebSocketClient` is updated with the credentials of the new user
- the current content of the DB is flushed (but the instance itself stays, so all existing controllers and workers are still alive and keep working)
- a new token is set (either directly, or from `config.tokenProvider`)
- `WebSocketClient.connect()` is called
- When the connection is established, the data about the current user are stored in the DB
- ... and finally, the completion block of `setCurrentUser` is called 🎉

### CODE REVIEW

File: StreamChatClient_v3/ChatClient.swift [hunk 1]
```diff
@@ -42,27 +42,21 @@ public typealias ChatClient = Client<DefaultDataTypes>
...
+    /// The id of the currently logged in user.
+    @Atomic public var currentUserId: UserId = .anonymous
     
-    /// The currently logged-in user.
-    public let currentUser: UserModel<ExtraData.User>
+    /// The currently logged-in user data.
+    ///
+    /// Returns `nil` if the user is not fully logged-in yet. Always wait for the `setUser` completion block before
+    /// accessing this value.
+    public var currentUser: CurrentUserModel<ExtraData.User>? {
```
Comment (buh) [inline StreamChatClient_v3/ChatClient.swift:52]: 1. It should be `internal(set)`
2. I think it shouldn't be optional, because we have anonymous by default as placeholder. It will simplify a lot of our lives.

File: StreamChatClient_v3/ChatClient.swift [hunk 2]
```diff
@@ -73,64 +67,61 @@ public class Client<ExtraData: ExtraDataTypes> {
     private(set) var backgroundWorkers: [Worker]!
     
     /// The `APIClient` instance `Client` uses to communicate with Stream REST API.
-    private(set) lazy var apiClient: APIClient = {
-        let apiClient = self.environment
-            .apiClientBuilder(self.config.apiKey, config.baseURL.restAPIBaseURL, self.urlSessionConfiguration)
-        self.webSocketClient.connectionStateDelegate = apiClient
+    lazy var apiClient: APIClient = {
```
Comment (buh) [inline StreamChatClient_v3/ChatClient.swift:70]: Add `private(set)` and the same for `webSocketClient`

File: StreamChatClient_v3/ChatClient.swift [hunk 3]
```diff
@@ -73,64 +67,61 @@ public class Client<ExtraData: ExtraDataTypes> {
...
+        // Create a connection request
+        let webSocketEndpoint = webSocketConnectEndpoint(userId: self.currentUserId)
+        
+        var encoder = environment.requestEncoderBuilder(config.baseURL.webSocketBaseURL, config.apiKey)
+        encoder.connectionDetailsProviderDelegate = self
+        
+        let connectEndpoint = webSocketConnectEndpoint(userId: currentUserId)
+        
+        // TODO: Add more middlewares
+        let eventMiddlewares = [
```
Comment (buh) [inline StreamChatClient_v3/ChatClient.swift:91]: If the `webSocketClient` is lazy var we can move this outside for easier management

File: StreamChatClient_v3/ChatClient.swift [hunk 4]
```diff
@@ -73,64 +67,61 @@ public class Client<ExtraData: ExtraDataTypes> {
...
-        }
+        return wsClient
     }()
     
     /// The `DatabaseContainer` instance `Client` uses to store and cache data.
-    private(set) lazy var persistentContainer: DatabaseContainer = {
+    private(set) lazy var databaseContainer: DatabaseContainer = {
         do {
             if config.isLocalStorageEnabled {
                 guard let storeURL = config.localStorageFolderURL else {
```
Comment (buh) [inline StreamChatClient_v3/ChatClient.swift:110]: I think the `storeURL` is related to the current user, or not?

File: StreamChatClient_v3/ChatClient.swift [hunk 5]
```diff
@@ -73,64 +67,61 @@ public class Client<ExtraData: ExtraDataTypes> {
...
                                                          attributes: nil)
-                let dbFileURL = config.localStorageFolderURL!.appendingPathComponent(currentUser.id)
+                
+                let dbFileURL = config.localStorageFolderURL!.appendingPathComponent(config.apiKey.apiKeyString)
                 return try environment.databaseContainerBuilder(.onDisk(databaseFileURL: dbFileURL))
             }
             
-        } catch let error as ClientError.MissingLocalStorageURL {
-            log.assert(false, "The URL provided in ChatClientConfig can't be `nil`.")
+        } catch is ClientError.MissingLocalStorageURL {
```
Comment (buh) [inline StreamChatClient_v3/ChatClient.swift:123]: I didn't know that 👍

File: StreamChatClient_v3/ChatClient.swift [hunk 6]
```diff
@@ -151,80 +142,264 @@ public class Client<ExtraData: ExtraDataTypes> {
...
     
-    /// Stream-specific request auth headers.
-    private var authSessionHeaders: [String: String] {
+    /// Stream-specific request headers.
+    private var sessionHeaders: [String: String] {
         var headers = [
             "X-Stream-Client": "stream-chat-swift-client-\(SystemEnvironment.version)",
             "X-Stream-Device": SystemEnvironment.deviceModelName,
             "X-Stream-OS": SystemEnvironment.systemName,
             "X-Stream-App-Environment": SystemEnvironment.name
```
Comment (buh) [inline StreamChatClient_v3/ChatClient.swift:155]: This is not really relevant anymore. Backend don't use additional headers. We should change it in this way:
```
"X-Stream-Client": "stream-chat-swift-client-\(SystemEnvironment.version)|device=\(SystemEnvironment.deviceModelName)|os=\(SystemEnvironment.systemName)..."
```

File: StreamChatClient_v3/ChatClient.swift [hunk 7]
```diff
@@ -151,80 +142,264 @@ public class Client<ExtraData: ExtraDataTypes> {
...
+                }
+            }
+        }
+    }
+    
+    /// Connects `Client` to the chat servers. When the connection is established, `Client` starts receiving chat updates.
+    ///
+    /// - Parameter completion: Called when the connection is established. If the connection fails, the completion is
+    /// called with an error.
+    public func connect(completion: ((Error?) -> Void)? = nil) {
```
Comment (buh) [inline StreamChatClient_v3/ChatClient.swift:298]: This is shouldn't be public. 
We should only connect on `setUser`, `setAnonymousUser` and etc

Comment (buh) [reply StreamChatClient_v3/ChatClient.swift:298]: When developers set `setUser` then after that we should keep connection and reconnect if went from background. 
We should stop do that if developers call `disconnect()` deliberately.

File: StreamChatClient_v3/Database/DatabaseContainer.swift [hunk 8]
```diff
@@ -107,6 +107,32 @@ class DatabaseContainer: NSPersistentContainer {
...
+    
+    /// Removes all data from the local storage.
+    ///
+    /// - Warning: ⚠️ This is a non-recoverable operation. All data will be lost after calling this method.
+    ///
+    /// - Parameters:
+    ///   - force: If sets to `false`, the method fails if there are unsynced data in to local storage, for example
+    /// messages pedning sent. You can use this option to warn a user about potential data loss.
+    ///   - completion: Called when the operation is completed. If the error is present, the operation failed.
+    func flush(force: Bool, completion: ((Error?) -> Void)? = nil) {
```
Comment (buh) [inline StreamChatClient_v3/Database/DatabaseContainer.swift:119]: > the current content of the DB is flushed (but the instance itself stays, so all existing controllers and workers are still alive and keep working)
I don't think we should do that. We discussed that user switch will switch to different database file source. So we can keep offline data for different users. Of course this is a rare operation from users, but why we should remove?
What's more important that user will connect on `setUser` with empty database if developers will do that because bad flow.

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift [hunk 9]
```diff
@@ -87,16 +98,25 @@ class WebSocketClient {
...
     
-    init(
-        urlRequest: URLRequest,
-        sessionConfiguration: URLSessionConfiguration,
-        eventDecoder: AnyEventDecoder,
-        eventMiddlewares: [EventMiddleware],
-        reconnectionStrategy: WebSocketClientReconnectionStrategy = DefaultReconnectionStrategy(),
-        environment: Environment = .init()
-    ) {
+    private func engineBuilder() throws -> WebSocketEngine {
```
Comment (buh) [inline StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:101]: It's not a builder. It returns `WebSocketEngine`

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift [hunk 10]
```diff
@@ -212,18 +241,22 @@ extension WebSocketClient: WebSocketEngineDelegate {
...
-            }
-            
             middlewares.process(event: event) { [weak self] event in
                 guard let self = self, let event = event else { return }
-                self.notificationCenter.post(Notification(newEventReceived: event, sender: self))
+                
+                if let event = event as? HealthCheck {
+                    if self.connectionState.isConnected == false {
+                        self.connectionState = .connected(connectionId: event.connectionId)
+                        self.reconnectionStrategy.sucessfullyConnected()
```
Comment (buh) [inline StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:250]: You can do that in `didSet` for `connectionState`

File: StreamChatClient_v3/ChatClient.swift (hunk 1 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/ChatClient.swift:52]: 1. It's a computed variable, you can't mutate it.

Comment (VojtaStavik) [reply StreamChatClient_v3/ChatClient.swift:52]: 2. I see what you mean but I think this will be a smaller issue in v3. Since the solution with an optional user is more correct, I'd suggest keeping it like this for now, and see how big of a complication it is for us. As I said, I'm open to changing it but I just want to wait till we see it's really a problem.

File: StreamChatClient_v3/ChatClient.swift (hunk 3 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/ChatClient.swift:91]: Hmm, OK, why not 🤷‍♂️

File: StreamChatClient_v3/ChatClient.swift (hunk 7 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/ChatClient.swift:298]: If we have `disconnect()` public, we need to have `connect()` public, too. Is there any reason why to *not* expose this API?

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift (hunk 9 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:101]: Not sure what you mean here. It's called `builder` to signalize that it creates a new instance, that it's not some existing object.

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift (hunk 10 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:250]: What do you think are the benefits of doing it there instead of doing it here?

File: StreamChatClient_v3/ChatClient.swift (hunk 6 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/ChatClient.swift:155]: So why do we still use it in v2? 🙄  @buh, where can I find more info about this?

File: StreamChatClient_v3/ChatClient.swift (hunk 1 above)
Comment (buh) [reply StreamChatClient_v3/ChatClient.swift:52]: I did the current user as optional at the beginning. But it was with a lot of noise to unwrap the current user. Additionally all our API will work only with logged in user after devs setup it. 
You can keep it. I recommend to think about that a little bit more

File: StreamChatClient_v3/ChatClient.swift (hunk 6 above)
Comment (buh) [reply StreamChatClient_v3/ChatClient.swift:155]: Unfortunately you have to ask Marcelo directly about additional headers.
I planned to change that in v2, because now it's useless.

File: StreamChatClient_v3/ChatClient.swift (hunk 7 above)
Comment (buh) [reply StreamChatClient_v3/ChatClient.swift:298]: Because we have to connect on `setUser` and keep connection as much as possible.
`disconnect` basically will logout the user.

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift (hunk 9 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:101]: So it doesn't return a builder it _creates_ an instance, then it should be renamed to `createEngine` as you do in other places.

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift (hunk 10 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:250]: I think it will be solid, because changing connection state should sometimes do a side effect and `didSet` is a best place for that. It was designed for that. We can be sure if the code will changed in future the behaviour will be the same. I think we can avoid mistakes.

File: StreamChatClient_v3/ChatClient.swift (hunk 6 above)
Comment (buh) [reply StreamChatClient_v3/ChatClient.swift:155]: I got confirmation from Marcel to do in this way:
`"X-Stream-Client": "stream-chat-swift-client-1.0.0|iPhone10|iOS13.6|production"`

File: StreamChatClient_v3/Config/ChatClientConfig.swift [hunk 11]
```diff
@@ -8,7 +8,7 @@ import Foundation
 ///
 /// The default configuration can be changed the following way:
 ///   ```
-///     var config = ChatClient.Config()
+///     var config = ChatClientConfig()
```
Comment (b-onc) [inline StreamChatClient_v3/Config/ChatClientConfig.swift:11]: Why did you move the config out of ChatClient?

File: StreamChatClient_v3/APIClient/RequestEncoder.swift [hunk 12]
```diff
@@ -191,8 +201,9 @@ private extension URL {
     }
 }
 
-protocol ConnectionIdProviderDelegate: AnyObject {
+protocol ConnectionDetailsProviderDelegate: AnyObject {
     func provideConnectionId(completion: @escaping (_ connectionId: ConnectionId?) -> Void)
+    func provideToken() -> Token?
```
Comment (b-onc) [inline StreamChatClient_v3/APIClient/RequestEncoder.swift:206]: Why isn't this async? 🤯

File: StreamChatClient_v3/WebSocketClient/WebSocketPayload.swift [hunk 13]
```diff
@@ -4,30 +4,44 @@
 
 import Foundation
 
-struct WebSocketPayload<ExtraData: ExtraDataTypes>: Encodable {
+struct WebSocketPayload<ExtraData: UserExtraData>: Encodable {
```
Comment (b-onc) [inline StreamChatClient_v3/WebSocketClient/WebSocketPayload.swift:7]: We need a better name for this, since different websocket payloads will contain different entities

File: StreamChatClient_v3/Database/DatabaseContainer.swift (hunk 8 above)
Comment (b-onc) [inline StreamChatClient_v3/Database/DatabaseContainer.swift:119]: The name `flush` implies something not completed is being forced to complete (ie flushing the output to the console), I suggest `nuke` for this 😂

File: StreamChatClient_v3/ChatClient.swift [hunk 14]
```diff
@@ -151,80 +142,264 @@ public class Client<ExtraData: ExtraDataTypes> {
...
+    /// is called with an error.
+    public func setAnonymousUser(shouldConnectAutomatically: Bool = true, completion: ((Error?) -> Void)? = nil) {
+        disconnect()
+        prepareEnvironmentForNewUser(userId: .anonymous, role: .anonymous, extraData: nil) { error in
+            guard error == nil else {
+                completion?(error)
+                return
+            }
+            
+            self.connect(completion: completion)
```
Comment (b-onc) [inline StreamChatClient_v3/ChatClient.swift:228]: `shouldConnectAutomatically` ignored?

File: StreamChatClient_v3/ChatClient.swift [hunk 15]
```diff
@@ -151,80 +142,264 @@ public class Client<ExtraData: ExtraDataTypes> {
...
+        disconnect()
+        
+        prepareEnvironmentForNewUser(userId: userId, role: .user, extraData: userExtraData) { error in
+            guard error == nil else {
+                completion?(error)
+                return
+            }
+            
+            if let token = token {
+                self.currentToken = token
```
Comment (b-onc) [inline StreamChatClient_v3/ChatClient.swift:277]: Why don't we:
```swift
if let token = token {
  config.tokenProvider = { $2(token) }
}
```
and treat the rest of the code as if only tokenProvider is provided?

File: StreamChatClient_v3/ChatClient.swift (hunk 7 above)
Comment (b-onc) [reply StreamChatClient_v3/ChatClient.swift:298]: I'm for giving users more control, login/logout is restrictred. I prefer login/connect/disconnect/logout as it's more granular

File: StreamChatClient_v3/APIClient/Endpoints/ChannelEndpoints.swift [hunk 16]
```diff
@@ -45,7 +45,7 @@ struct ChannelPayload<ExtraData: ExtraDataTypes>: Decodable {
     init(from decoder: Decoder) throws {
         let container = try decoder.container(keyedBy: CodingKeys.self)
         channel = try container.decode(ChannelDetailPayload<ExtraData>.self, forKey: .channel)
-        watcherCount = try container.decode(Int.self, forKey: .watcherCount)
+        watcherCount = try container.decodeIfPresent(Int.self, forKey: .watcherCount) ?? 0
```
Comment (b-onc) [inline StreamChatClient_v3/APIClient/Endpoints/ChannelEndpoints.swift:48]: I also have this change on my PR, I think `watcherCount` should be optional

File: StreamChatClient_v3/Config/ChatClientConfig.swift (hunk 11 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/Config/ChatClientConfig.swift:11]: This is just a comment which wasn't updated. It was wrong. We have `Config` outside of Client since the time we made `Client` generic :) You don't want to write `Client<ExtraData>.Config`,

File: StreamChatClient_v3/ChatClient.swift (hunk 1 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/ChatClient.swift:52]: As I said, I'm not against it, and you're probably right. I just try to keep it the most correct way as long as possible/convenient. Once it gets annoying, we change this.

File: StreamChatClient_v3/ChatClient.swift (hunk 4 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/ChatClient.swift:110]: Discussed in Slack

File: StreamChatClient_v3/Database/DatabaseContainer.swift (hunk 8 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/Database/DatabaseContainer.swift:119]: Discussed in Slack

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift (hunk 9 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:101]: It's actually `builder` everywhere, but if you prefer `createEngine` I have no problem with that.

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift (hunk 10 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:250]: Ok, fair enough 👍

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift [hunk 17]
```diff
@@ -45,14 +45,26 @@ class WebSocketClient {
...
-    private lazy var engine: WebSocketEngine = {
-        let engine = self.environment.engineBuilder(self.urlRequest, self.sessionConfiguration, self.engineQueue)
-        engine.delegate = self
-        return engine
+    lazy var engine: WebSocketEngine = {
+        do {
+            return try engineBuilder()
+        } catch {
+            fatalError("Failed to create WebSocketEngine with error: \(error)")
+        }
```
[the remark this answers is not in the document]
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:67]: Yep, we can do that.

File: StreamChatClient_v3/Database/DTOs/CurrentUserDTO.swift [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+extension CurrentUserModel {
+    static func create(fromDTO dto: CurrentUserDTO) -> CurrentUserModel {
+        let user = dto.user
+        
+        let extraData: ExtraData
+        do {
+            extraData = try JSONDecoder.default.decode(ExtraData.self, from: user.extraData)
+        } catch {
+            fatalError("Failed decoding saved extra data with error: \(error). This should never happen because"
+                + "the extra data must be a valid JSON to be saved.")
```
[the remark this answers is not in the document]
Comment (VojtaStavik) [reply StreamChatClient_v3/Database/DTOs/CurrentUserDTO.swift:70]: This was a courtesy and a note to future us when we encounter this impossible error :)

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift (hunk 9 above)
Comment (buh) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:101]: The function named `engineBuilder()` and I think it's clear that I expect it will return a builder and then we can use the builder to create engines. But the content of the function is different. It creates an engine.
You on the same time in other places user prefix `create` for cases like that and it's not clear why you decide to change it here.

File: StreamChatClient_v3/APIClient/RequestEncoder.swift (hunk 12 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/APIClient/RequestEncoder.swift:206]: In the current setup, it doesn't need to be. Maybe it will have to be async once we implement the token refresh functionality, but I didn't want to complicate things now.

File: StreamChatClient_v3/WebSocketClient/WebSocketPayload.swift (hunk 13 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketPayload.swift:7]: This is actually the outgoing payload. It's the initial payload we open the WS connection with. It won't contain different entities. This is it. Any idea for a name?

Comment (b-onc) [reply StreamChatClient_v3/WebSocketClient/WebSocketPayload.swift:7]: `WebSocketConnectPayload`

File: StreamChatClient_v3/Database/DatabaseContainer.swift (hunk 8 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/Database/DatabaseContainer.swift:119]: Yes, the naming is not correct there. I'll replace it with `deleteAllData`, which is much better and more descriptive.

File: StreamChatClient_v3/ChatClient.swift (hunk 14 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/ChatClient.swift:228]: Haha, I decided to remove it and forgot to remove it from the signature :) Forget about `shouldConnectAutomatically`...

File: StreamChatClient_v3/ChatClient.swift (hunk 15 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/ChatClient.swift:277]: Hmm, the current setup allows you to override the provider with a token if you need it. I like your suggestion for its cleverness but I'd stay to the more boring implementation for this :)

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift (hunk 10 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift]: fixed in 42bc4ba25099008d0f338e13d3bc4b769a6688d1

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift (hunk 17 above)
[the remark this answers is not in the document]
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift]: fixed in a1845ca42ed2ef4f20a719e103bfe25c3b092a86

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift [hunk 19]
```diff
@@ -45,14 +45,26 @@ class WebSocketClient {
     /// Event middlewares used to pre-process incoming events before they are published
     var middlewares: [EventMiddleware]
     
+    /// The endpoint used for creating a web socket connection.
+    ///
+    /// Changing this value doesn't automatically update the existing connection. You need to manually call `disconnect`
+    /// and `connect` to make a new connection to the updated endpoint.
+    var endpoint: Endpoint<EmptyResponse> {
```
[the remark this answers is not in the document]
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift]: fixed in 6cce89176a804052631ed70d877185344f2fec58 and 3664d4a2a599d8c443813e19a57911cb96a14c67

File: StreamChatClient_v3/APIClient/Endpoints/ChannelEndpoints.swift (hunk 16 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/APIClient/Endpoints/ChannelEndpoints.swift]: OK, removing this from my PR

File: StreamChatClient_v3/Database/DatabaseContainer.swift (hunk 8 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/Database/DatabaseContainer.swift]: fixed

File: StreamChatClient_v3/WebSocketClient/WebSocketPayload.swift (hunk 13 above)
Comment (b-onc) [reply StreamChatClient_v3/WebSocketClient/WebSocketPayload.swift:7]: ⬆️ suggestion for a name

File: StreamChatClient_v3/ChatClient.swift (hunk 6 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/ChatClient.swift]: please check 1236cc92a9f211ca970a451c985790c0e52595e9

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift (hunk 9 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/WebSocketClient/WebSocketClient.swift]: fixed

File: StreamChatClient_v3/ChatClient.swift (hunk 14 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/ChatClient.swift:219]: fixed

File: StreamChatClient_v3/ChatClient.swift (hunk 6 above)
Comment (buh) [reply StreamChatClient_v3/ChatClient.swift]: 1. Actually as I understand no need labels. You can pass just values splitter by `|`
2. I think it's good to add `SystemEnvironment.name` too

File: StreamChatClient_v3/ChatClient.swift (hunk 7 above)
Comment (buh) [reply StreamChatClient_v3/ChatClient.swift:289]: We discussed that in the past. Connect should happened on setUser.

File: StreamChatClient_v3/WebSocketClient/WebSocketClient.swift [hunk 20]
```diff
@@ -87,16 +94,30 @@ class WebSocketClient {
...
     
-    init(
-        urlRequest: URLRequest,
-        sessionConfiguration: URLSessionConfiguration,
-        eventDecoder: AnyEventDecoder,
-        eventMiddlewares: [EventMiddleware],
-        reconnectionStrategy: WebSocketClientReconnectionStrategy = DefaultReconnectionStrategy(),
-        environment: Environment = .init()
-    ) {
+    private func engineBuilder() -> WebSocketEngine {
```
Comment (buh) [inline StreamChatClient_v3/WebSocketClient/WebSocketClient.swift:97]: Is it still a builder? 😄

File: StreamChatClient_v3/ChatClient.swift (hunk 7 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/ChatClient.swift:289]: It does. Please check the implementation, or the flow described in the PR description. `connect` is called on `setUser` automatically. There's no reason, why users shouldn't be able to call `disconnect()` and `connect()` if they want :) Especially, when we have this functionality implemented.

File: StreamChatClient_v3/ChatClient.swift (hunk 6 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/ChatClient.swift]: I removed the labels and added `name`
